text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+- Feature Name: target_bundles
+- Start Date: 2016-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Combine distribution of standard libraries and targets into bundles for targeting a particular
+platform. Such bundle is all you need from the Rust side to cross-compile for your target.
+
+The major ideas of this RFC is to:
+
+1. Make JSON targets as full featured as they need to be in order to support specifying targets as
+   custom as they need to be;
+2. Convert current built-in targets to the JSON targets;
+3. Change distribution of libstd to include a corresponding JSON target.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there’s two different ways rustc targets are distributed: built-in targets and custom
+JSON targets. Built-in targets are very inflexible – they cannot be changed without changing and
+recompiling the compiler itself. On the other side of the spectrum, custom JSON targets are easy to
+adjust and adapt, but feature wise are very limited and rarely are suitable for the more uncommon
+use-cases.
+
+We’ve observed a considerable amount of desire by users of the language to customize targets they
```

Comment:
A hash of the target spec makes sense to me. I think it would have to be appended to the target triple of the rustlib directories: `$sysroot/lib/rustlib/$TARGET-$HASH`. That way several variants of the same target triple can coexist in the sysroot. It may also be worth to encode the hash in the rlibs' metadata for "extra protection"; someone might rename a rustlib directory from `$TARGET-$HASH1` to `$TARGET-$HASH2` -- the hash in metadata would prevent using the wrong variant in that case.